src/context/product-context.js

```diff
@@ -1,28 +1,65 @@
-import { createContext, useContext, useEffect, useState } from "react";
+import { createContext, useContext, useEffect, useReducer } from "react";
 import axios from "axios";
+import productReducer from "../reducer/product-reducer";
 
 const ProductContext = createContext();
 
+const products = []; // list of all products
+const filteredProducts = []; // list of products that will be displayed
+const noProducts = 0; // number of products currently displayed
+const categories = [
+  // categories with filter status
+  { name: "perennial", isChecked: true },
+  { name: "palm", isChecked: true },
+  { name: "indoor", isChecked: true },
+  { name: "fruit", isChecked: true },
+];
+const rating = [
+  // ratings with filter status
+  { stars: 4, isChecked: false },
+  { stars: 3, isChecked: false },
+  { stars: 2, isChecked: false },
+  { stars: 1, isChecked: true },
+];
+const range = 500; // max price of products
+
 export function ProductProvider({ children }) {
-  const [products, setProducts] = useState([]);
+  const [productState, productDispatch] = useReducer(productReducer, {
+    products,
+    filteredProducts,
+    noProducts,
+    categories,
+    rating,
+    range,
+  });
 
+  // API call to GET products
   async function getProducts() {
     let response = [];
     try {
       response = await axios.get("/api/products");
     } catch (error) {
-      console.log(error);
+      console.log(error); // todo: add error and loading feedback
     } finally {
-      setProducts(response.data.products);
+      productDispatch({
+        type: "ADD_PRODUCTS",
+        payload: response.data.products,
+      });
     }
   }
 
+  // called once in the beginning
   useEffect(() => {
     getProducts();
   }, []);
 
+  // updates the filteredProducts whenever a filter state changes
+  useEffect(() => {
+    productDispatch({ type: "APPLY_FILTERS" });
+  }, [productState.categories, productState.rating, productState.range]);
```

Comment:
Calling useEffect on every change of the filters and rerunning all the filters might be heavy. My approach would be you apply all the filters just before rendering the card and show the card with the suitable filter. do let me know what you think.

Reply:
Since the number of products was low, I decided to fetch them only once. This made the re-renders less costly. As the number of products increase, we can use pages like [these](https://shop4reebok.com/#c/men-products)  to limit the data fetched at once(let's say 20 products). This will hopefully make things easier for client side rendering as there will be only 20 products in the view at a time